src/TagInputStateful/index.jsx

```diff
@@ -0,0 +1,623 @@
+        onMouseOver       : undefined,
+        onMouseOverOption : undefined,
+        options           : undefined,
+        placeholder       : undefined,
+        tags              : undefined,
+        value             : '',
+    };
+
+    constructor( props )
+    {
+        super( props );
+
+        const dropdownPosition = props.dropdownPosition !== 'auto' ?
+            props.dropdownPosition : 'bottom';
+
+        this.state = {
+            activeOption    : undefined,
+            dropdownPosition,
+            filteredOptions : undefined,
+            isOpen          : props.isOpen,
+            options         : props.options,
+            selection       : undefined,
+            tags            : props.tags,
+            value           : props.value,
+        };
+
+        this.handleBlur            = this.handleBlur.bind( this );
+        this.handleChange          = this.handleChange.bind( this );
+        this.handleClickClose      = this.handleClickClose.bind( this );
+        this.handleClickOption     = this.handleClickOption.bind( this );
+        this.handleFocus           = this.handleFocus.bind( this );
+        this.handleKeyDown         = this.handleKeyDown.bind( this );
+        this.handleMouseOutOption  = this.handleMouseOutOption.bind( this );
+        this.handleMouseOverOption = this.handleMouseOverOption.bind( this );
+        this.setInputRef           = this.setInputRef.bind( this );
+        this.setScrollBoxRef       = this.setScrollBoxRef( this );
+        this.setWrapperRef         = this.setWrapperRef.bind( this );
+    }
+
+    componentDidMount()
+    {
+        this.setDropdownPosition();
+    }
+
+    componentWillReceiveProps( newProps )
+    {
+        this.setDropdownPosition( newProps );
+    }
+
+    componentDidUpdate()
+    {
+        const { scrollBox } = this;
+        const { activeOption, id } = this.props;
+
+        if ( scrollBox && activeOption )
+        {
+            const activeEl =
+                document.getElementById( addPrefix( activeOption, id ) );
+
+            if ( activeEl && scrollBox.scrollHeight > scrollBox.offsetHeight )
+            {
+                const pos        = activeEl.offsetTop;
+                const elHeight   = activeEl.offsetHeight;
+                const contHeight = scrollBox.offsetHeight;
+
+                const min = scrollBox.scrollTop;
+                const max = min + ( scrollBox.offsetHeight - elHeight );
+
+                if ( pos < min )
+                {
+                    scrollBox.scrollTop = pos;
+                }
+                else if ( pos > max )
+                {
+                    scrollBox.scrollTop = pos - ( contHeight - elHeight );
+                }
+            }
+        }
+    }
+
+    setInputRef( ref )
+    {
+        this.inputRef = ref;
+    }
+
+    setScrollBoxRef( ref )
+    {
+        if ( ref )
+        {
+            this.scrollBox = ref;
+        }
+    }
+
+    setWrapperRef( ref )
```

Comment:
For simplicity we should use `React.createRef()` rather than having to define these `set...Ref` functions.